program/rust/src/time_machine_types.rs

```diff
@@ -1,45 +1,192 @@
+    /// Numerator for the current epoch (ex : slot_gap1 * price_1 + slot_gap2 * price_2 + slot_gap3
+    /// * price_3)
+    /// It can never overflow because :
+    /// slot_gap1 * price_1 + slot_gap2 * price_2 + slot_gap3 * price_3 <= (slot_gap1 + slot_gap2 +
+    /// slot_gap3) * i64::MAX <= clock_slot * i128::MAX
+    pub current_epoch_numerator:       i128,
+    /// Denominator for the current epoch (ex : slot_gap1 + slot_gap2 + slot_gap3)
+    /// It can never overflow because : (slot_gap1 + slot_gap2 + slot_gap3) <= clock_slot
+    pub current_epoch_denominator:     u64,
+    /// Whether the current epoch is valid (i.e. max slot_gap <= threshold) for all the epochs
+    /// updates
+    pub current_epoch_is_valid:        bool,
+    /// Stores the running sum of individual epoch averages, we can support u64::MAX epochs. And
+    /// clock.unix_timestamp <= u64::MAX so the clock will break first.
+    pub running_sum_of_price_averages: [i128; NUM_BUCKETS],
+    /// Stores the number of valid epochs since inception, we can support u64::MAX epochs. And
+    /// clock.unix_timestamp <= u64::MAX so the clock will break first.
+    pub running_valid_epoch_counter:   [u64; NUM_BUCKETS],
+}
+
+impl<const NUM_BUCKETS: usize> SmaTracker<NUM_BUCKETS> {
+    pub fn time_to_epoch(&self, time: i64) -> Result<usize, OracleError> {
+        try_convert::<i64, usize>(time / self.granularity) // Can never fail because usize is u64
+                                                           // and time is positive
+    }
+
+    pub fn initialize(&mut self, granularity: i64, threshold: u64) {
+        self.threshold = threshold;
+        self.granularity = granularity;
+        self.current_epoch_is_valid = true;
+    }
+
+    pub fn add_datapoint(&mut self, datapoint: &DataPoint) -> Result<(), OracleError> {
+        let epoch_0 = self.time_to_epoch(datapoint.last_two_timestamps.0)?;
+        let epoch_1 = self.time_to_epoch(datapoint.last_two_timestamps.1)?;
+
+        let datapoint_numerator = i128::from(datapoint.slot_gap) * i128::from(datapoint.price);
+        let datapoint_denominator = datapoint.slot_gap;
+
+        self.current_epoch_denominator += datapoint_denominator;
+        self.current_epoch_numerator += datapoint_numerator;
+
+        self.current_epoch_is_valid =
+            self.current_epoch_is_valid && datapoint_denominator <= self.threshold;
+
+        // If epoch changed
+        if epoch_0 < epoch_1 {
+            let index = epoch_0.rem_euclid(NUM_BUCKETS);
+
+            // This addition can only overflow if NUM_BUCKETS ~ u64::MAX / 2
+            let prev_index = (epoch_0 + NUM_BUCKETS - 1).rem_euclid(NUM_BUCKETS);
+
+            self.running_sum_of_price_averages[index] = self.running_sum_of_price_averages
+                [prev_index]
+                + self.current_epoch_numerator / i128::from(self.current_epoch_denominator);
+            // The fraction here is smaller than i64::MAX , so we can support u64::MAX updates
```

Comment:
weird location for this comment. put it above the line of code that it applies to, not below.

Reply:
Yes